File: test/fungible-token.test.js
import {
    executeScript,
    getContractAddress,
    sendTransaction
} from "flow-js-testing";
import {
    addFlowTokensToDepositor,
    after,
    alice,
    before, cleanup, exampleNFTAdmin,
    exampleTokenAdmin, getEventFromTransaction,
    getRedeemableTypes,
    getTicketDepositFromRes,
    lostAndFoundAdmin
} from "./common";

// Increase timeout if your tests failing due to timeout
jest.setTimeout(10000);

const depositAmount = 1000

describe("lost-and-found FungibleToken tests", () => {
    beforeEach(async () => {
        await before()
    });

    // Stop emulator, so it could be restarted
    afterEach(async () => {
        await after()
    });

    const depositExampleToken = async (account, amount) => {
        const exampleTokenAddress = await getContractAddress("ExampleToken")

        const args = [account, amount]
        const signers = [exampleTokenAddress]
        let [tx, err] = await sendTransaction({ name: "ExampleToken/deposit_example_token", args, signers });
        return [tx, err]
    }

    const configureExampleToken = async (account) => {
        const signers = [account]
        let [tx, err] = await sendTransaction({ name: "ExampleToken/setup_vault", args: [], signers })
        return [tx, err]
    }

    test("deposit ExampleToken", async () => {
        const args = [alice, depositAmount]
        const signers = [exampleTokenAdmin]
        let [tx, err] = await sendTransaction({ name: "ExampleToken/deposit_example_token", args, signers });
        expect(err).toBe(null)

        let result
        [result, err] = await getRedeemableTypes(alice)
        expect(err).toBe(null)
        let found = false
        result.forEach(val => {
            if (val.typeID === `A.${exampleTokenAdmin.substring(2)}.ExampleToken.Vault`) {
                found = true
            }
        })
        expect(found).toBe(true)
        const eventType = `A.${lostAndFoundAdmin.substring(2)}.LostAndFound.TicketDeposited`
        const event = getEventFromTransaction(tx, eventType)
        const ticketID = event.data.ticketID
        expect(ticketID).toBeGreaterThan(0)
        let [ticketDetail, ticketDetailErr] = await executeScript("ExampleToken/borrow_ticket", [alice, ticketID])
        expect(ticketDetailErr).toBe(null)
        expect(ticketDetail.redeemer).toBe(alice)
        expect(ticketDetail.type.typeID).toBe(`A.${exampleTokenAdmin.substring(2)}.ExampleToken.Vault`)
    })

    test("redeem ExampleToken", async () => {
        await depositExampleToken(alice, depositAmount)
        let [_, redeemErr] = await sendTransaction({
            name: "ExampleToken/redeem_example_token_all",
            args: [],
            signers: [alice],
            limit: 9999
        })
        expect(redeemErr).toBe(null)

        let [balance, balanceErr] = await executeScript("ExampleToken/get_example_token_balance", [alice])
        expect(balanceErr).toBe(null)
        expect(Number(balance)).toBeGreaterThanOrEqual(depositAmount)

        const [redeemableTypes, rtErr] = await getRedeemableTypes(alice)
        expect(rtErr).toBe(null)
        let found = false
        redeemableTypes.forEach(val => {
            if (val.typeID === `A.${exampleTokenAdmin.substring(2)}.ExampleToken.Vault`) {
                found = true
            }
        })
        expect(found).toBe(false)
    })

    test("send ExampleToken with setup", async () => {
        let [_, setupErr] = await sendTransaction({
            name: "ExampleToken/setup_account_ft",
            args: [],
            signers: [alice],
            limit: 999
        })
        expect(setupErr).toBe(null)

        let [balance, balanceErr] = await executeScript("ExampleToken/get_example_token_balance", [alice])
        expect(balanceErr).toBe(null)

        let [sendRes, sendErr] = await sendTransaction({
            name: "ExampleToken/try_send_example_token",
            args: [alice, depositAmount],
            signers: [exampleTokenAdmin],
            limit: 9999
        })
        expect(sendErr).toBe(null)
        const eventType = `A.${exampleTokenAdmin.substring(2)}.ExampleToken.TokensDeposited`
        const event = getEventFromTransaction(sendRes, eventType)
        expect(event.type).toBe(eventType)
        expect(Number(event.data.amount)).toBe(depositAmount)
        expect(event.data.to).toBe(alice)

        let [balanceAfter, balanceAfterErr] = await executeScript("ExampleToken/get_example_token_balance", [alice])
        expect(balanceAfterErr).toBe(null)
        expect(Number(balanceAfter)).toBe(depositAmount + Number(balance))
    })

    test("send ExampleToken without setup", async () => {
        await sendTransaction({
            name: "ExampleToken/destroy_example_token_storage",
            signers: [alice],
            args: [],
            limit: 9999
        })

        let [balance, balanceErr] = await executeScript("ExampleToken/get_example_token_balance", [alice])
        expect(balanceErr.message.includes("unexpectedly found nil while forcing an Optional value")).toBe(true)

        let [sendRes, sendErr] = await sendTransaction({
            name: "ExampleToken/try_send_example_token",
            args: [alice, depositAmount],
            signers: [exampleTokenAdmin],
            limit: 9999
        })
        expect(sendErr).toBe(null)

        const eventType = `A.${lostAndFoundAdmin.substring(2)}.LostAndFound.TicketDeposited`
        const event = getEventFromTransaction(sendRes, eventType)
        expect(event !== null).toBe(true)
        expect(event.type).toBe(eventType)
        expect(event.data.type.typeID).toBe(`A.${exampleTokenAdmin.substring(2)}.ExampleToken.Vault`)
        expect(event.data.redeemer).toBe(alice)
    })

    it("should send ExampleToken through Depositor", async () => {
        await cleanup(alice)
        await addFlowTokensToDepositor(exampleTokenAdmin, 1)

        let [setupRes, setupErr] = await sendTransaction({
            name: "ExampleToken/setup_account_ft",
            args: [],
            signers: [alice],
            limit: 999
        })
        expect(setupErr).toBe(null)

        const tokensToSend = 100
        const [sendRes, sendErr] = await sendTransaction({
            name: "ExampleToken/try_send_example_token_depositor",
            args: [alice, tokensToSend],
            signers: [exampleTokenAdmin],
            limit: 9999
        })
        const eventType = `A.${exampleTokenAdmin.substring(2)}.ExampleToken.TokensDeposited`
        const event = getEventFromTransaction(sendRes, eventType)
        expect(sendErr).toBe(null)
        expect(event.type).toBe(eventType)
        expect(event.data.to).toBe(alice)

        let [balance, balanceErr] = await executeScript("ExampleToken/get_example_token_balance", [alice])
        expect(balanceErr).toBe(null)
    })

    it("should deposit ExampleToken through Depositor", async () => {
        await cleanup(alice)
        await addFlowTokensToDepositor(exampleTokenAdmin, 1)

        const tokensToSend = 100
        const [sendRes, sendErr] = await sendTransaction({
            name: "ExampleToken/try_send_example_token_depositor",
            args: [alice, tokensToSend],
            signers: [exampleTokenAdmin],
            limit: 9999
        })
        const eventType = `A.${lostAndFoundAdmin.substring(2)}.LostAndFound.TicketDeposited`
        const event = getEventFromTransaction(sendRes, eventType)
        expect(sendErr).toBe(null)
        expect(event.data.redeemer).toBe(alice)
        expect(event.data.type.typeID).toBe(`A.${exampleTokenAdmin.substring(2)}.ExampleToken.Vault`)
    })

    it("should be able to get total vaults balance to a specific bin", async () => {
        await cleanup(alice)
        await addFlowTokensToDepositor(exampleTokenAdmin, 1)

        const tokensToSend = 100.00000000
        const tokenType = `A.${exampleTokenAdmin.substring(2)}.ExampleToken.Vault`
        const [sendRes, sendErr] = await sendTransaction({
            name: "ExampleToken/try_send_example_token_depositor",
            args: [alice, tokensToSend],
            signers: [exampleTokenAdmin],
            limit: 9999
        })
        const eventType = `A.${lostAndFoundAdmin.substring(2)}.LostAndFound.TicketDeposited`
        const event = getEventFromTransaction(sendRes, eventType)
        expect(event.data.redeemer).toBe(alice)
        expect(event.data.type.typeID).toBe(`A.${exampleTokenAdmin.substring(2)}.ExampleToken.Vault`)

        let [balance, balanceErr] = await executeScript("ExampleToken/get_bin_vault_balance", [alice, tokenType])
        expect(balanceErr).toBe(null)
        expect(balance).toBe("100.00000000")

        const [sendRes2, sendErr2] = await sendTransaction({
            name: "ExampleToken/try_send_example_token_depositor",
            args: [alice, tokensToSend],
            signers: [exampleTokenAdmin],
            limit: 9999
        })
        expect(sendErr2).toBe(null)

        let [balance2, balanceErr2] = await executeScript("ExampleToken/get_bin_vault_balance", [alice, tokenType])
        expect(balanceErr2).toBe(null)
        expect(balance2).toBe("200.00000000")

    })

    it("should get the flow repayment address from script", async () => {
        await cleanup(alice)
        await addFlowTokensToDepositor(exampleTokenAdmin, 1)

        const tokensToSend = 100.00000000
        const tokenType = `A.${exampleTokenAdmin.substring(2)}.ExampleToken.Vault`
        const [sendRes, sendErr] = await sendTransaction({
            name: "ExampleToken/try_send_example_token_depositor",
            args: [alice, tokensToSend],
            signers: [exampleTokenAdmin],
            limit: 9999
        })
        const eventType = `A.${lostAndFoundAdmin.substring(2)}.LostAndFound.TicketDeposited`
        const event = getEventFromTransaction(sendRes, eventType)
        expect(event.data.redeemer).toBe(alice)
        expect(event.data.type.typeID).toBe(`A.${exampleTokenAdmin.substring(2)}.ExampleToken.Vault`)

        const ticketID = event.data.ticketID

        let [repaymentAddress, repaymentAddressErr] = await executeScript("get_flowRepayment_address", [alice, tokenType, ticketID])
        expect(repaymentAddressErr).toBe(null)
        expect(repaymentAddress).toBe(`${exampleTokenAdmin}`)

    })
})

File: fungible_token_test.go
package test_main

import (
	"testing"

	. "github.com/bjartek/overflow"
	"github.com/hexops/autogold"
	"github.com/stretchr/testify/assert"
)

func TestFungibleToken(t *testing.T) {

	otu := NewOverflowTest(t)
	depositAmount := 1000.0

	t.Run("deposit ExampleToken", func(t *testing.T) {

		user := "user1"

		TxWithArgs := otu.O.TxFN(
			WithArg("redeemer", user),
			WithArg("amount", depositAmount),
		)

		signer := WithSigner("account")

		ticketID, err := TxWithArgs("ExampleToken/deposit_example_token",
			signer,
		).
			AssertEvent(t, "TicketDeposited", map[string]interface{}{
				"type": "A.f8d6e0586b0a20c7.ExampleToken.Vault",
			}).
			GetIdFromEvent("TicketDeposited", "ticketID")

		assert.NoError(t, err)

		assert.Greater(t, ticketID, uint64(0))

		otu.O.Script("ExampleToken/borrow_ticket_as_struct",
			WithArg("addr", "user1"),
			WithArg("ticketID", ticketID),
		).
			AssertWithPointer(t, "/redeemer", otu.O.Address("user1")).
			AssertWithPointer(t, "/type", "A.f8d6e0586b0a20c7.ExampleToken.Vault")

	})

	t.Run("redeem ExampleToken", func(t *testing.T) {

		otu.depositExampleToken("user1", depositAmount)

		otu.O.Tx("ExampleToken/redeem_example_token_all",
			WithSigner("user1"),
		).
			AssertSuccess(t)

		balance, err := otu.O.Script("ExampleToken/get_example_token_balance",
			WithArg("addr", "user1"),
		).
			GetAsInterface()

		assert.NoError(t, err)

		assert.GreaterOrEqual(t, balance.(float64), depositAmount)

		// Shouldn't exist in redeemable type anymore
		otu.O.Script("get_redeemable_types_for_addr",
			WithArg("addr", "user1"),
		).Print().
			AssertWant(t, autogold.Want("This should be nil", nil))

	})

	t.Run("send ExampleToken with setup", func(t *testing.T) {

		otu.O.Tx("ExampleToken/setup_account_ft",
			WithSigner("user1"),
		).
			AssertSuccess(t)

		balance, err := otu.O.Script("ExampleToken/get_example_token_balance",
			WithArg("addr", "user1"),
		).
			GetAsInterface()

		assert.NoError(t, err)

		otu.O.Tx("ExampleToken/try_send_example_token",
			WithSigner("account"),
			WithArg("recipient", "user1"),
			WithArg("amount", depositAmount),
		).
			AssertSuccess(t).
			AssertEvent(t, "ExampleToken.TokensDeposited", map[string]interface{}{
				"amount": depositAmount,
				"to":     otu.O.Address("user1"),
			})

		otu.O.Script("ExampleToken/get_example_token_balance",
			WithArg("addr", "user1"),
		).
			AssertWant(t, autogold.Want("balance should be equal to amount + balance", depositAmount+balance.(float64)))

	})

	t.Run("send ExampleToken without setup", func(t *testing.T) {

		otu.O.Tx("ExampleToken/destroy_example_token_storage",
			WithSigner("user1"),
		).
			AssertSuccess(t)

		_, err := otu.O.Script("ExampleToken/get_example_token_balance",
			WithArg("addr", "user1"),
		).
			GetAsInterface()

		assert.Error(t, err)
		if err != nil {
			assert.Contains(t, err.Error(), "unexpectedly found nil while forcing an Optional value")
		}

		otu.O.Tx("ExampleToken/try_send_example_token",
			WithSigner("account"),
			WithArg("recipient", "user1"),
			WithArg("amount", depositAmount),
		).
			AssertSuccess(t).
			AssertEvent(t, "TicketDeposited", map[string]interface{}{
				"type":     "A.f8d6e0586b0a20c7.ExampleToken.Vault",
				"redeemer": otu.O.Address("user1"),
			})

	})

	t.Run("should send ExampleToken through Depositor", func(t *testing.T) {
		otu.cleanup("user1").
			setupDepositor("account", 0.0).
			addFlowToDepositor("account", 1.0).
			assertDepositorBalance("account", 1.0)

		otu.O.Tx("ExampleToken/setup_account_ft",
			WithSigner("user1"),
		).
			AssertSuccess(t)

		tokensToSend := 100.0

		otu.O.Tx("ExampleToken/try_send_example_token",
			WithSigner("account"),
			WithArg("recipient", "user1"),
			WithArg("amount", tokensToSend),
		).
			AssertSuccess(t).
			AssertEvent(t, "ExampleToken.TokensDeposited", map[string]interface{}{
				"amount": tokensToSend,
				"to":     otu.O.Address("user1"),
			})

		// assert 0 balance
		otu.O.Script("ExampleToken/get_example_token_balance",
			WithArg("addr", "user1"),
		).
			AssertWant(t, autogold.Want("should be equal to amount sent in", tokensToSend))

	})

	t.Run("should deposit ExampleToken through Depositor", func(t *testing.T) {
		otu.cleanup("user1")

		tokensToSend := 100.0

		otu.O.Tx("ExampleToken/try_send_example_token",
			WithSigner("account"),
			WithArg("recipient", "user1"),
			WithArg("amount", tokensToSend),
		).
			AssertSuccess(t).
			AssertEvent(t, "TicketDeposited", map[string]interface{}{
				"type":     "A.f8d6e0586b0a20c7.ExampleToken.Vault",
				"redeemer": otu.O.Address("user1"),
			})

	})

	ticketID := uint64(0)

	// New script test
	t.Run("should be able to get total vaults balance to a specific bin", func(t *testing.T) {
		otu.cleanup("user1")

		tokensToSend := 100.0

		// Send in 100.0 token to user 1's Lost and Found bin
		otu.O.Tx("ExampleToken/try_send_example_token",
			WithSigner("account"),
			WithArg("recipient", "user1"),
			WithArg("amount", tokensToSend),
		).
			AssertSuccess(t).
			AssertEvent(t, "TicketDeposited", map[string]interface{}{
				"type":     "A.f8d6e0586b0a20c7.ExampleToken.Vault",
				"redeemer": otu.O.Address("user1"),
			})

		// query the bin's balance. Assert it to be 100
		otu.O.Script("ExampleToken/get_bin_vault_balance",
			WithArg("addr", "user1"),
			WithArg("type", "A.f8d6e0586b0a20c7.ExampleToken.Vault"),
		).
			AssertWant(t, autogold.Want("should be equal to total sum 100", tokensToSend))

		// Send in 100.0 more token to user 1's Lost and Found bin
		tID, err := otu.O.Tx("ExampleToken/try_send_example_token",
			WithSigner("account"),
			WithArg("recipient", "user1"),
			WithArg("amount", tokensToSend),
		).
			AssertSuccess(t).
			AssertEvent(t, "TicketDeposited", map[string]interface{}{
				"type":     "A.f8d6e0586b0a20c7.ExampleToken.Vault",
				"redeemer": otu.O.Address("user1"),
			}).
			GetIdFromEvent("TicketDeposited", "ticketID")

		assert.NoError(t, err)

		ticketID = tID

		// query the bin's balance. Assert it to be 200
		otu.O.Script("ExampleToken/get_bin_vault_balance",
			WithArg("addr", "user1"),
			WithArg("type", "A.f8d6e0586b0a20c7.ExampleToken.Vault"),
		).
			AssertWant(t, autogold.Want("should be equal to total sum 200", tokensToSend*2))

	})

	t.Run("should get the flow repayment address from script", func(t *testing.T) {

		// query the bin's balance. Assert it to be 100
		otu.O.Script("get_flowRepayment_address",
			WithArg("addr", "user1"),
			WithArg("type", "A.f8d6e0586b0a20c7.ExampleToken.Vault"),
			WithArg("ticketID", ticketID),
		).
			AssertWant(t, autogold.Want("should get admin account address", otu.O.Address("account")))

	})

}
